api/client/websocketupgradeproto/protocol.go

```diff
@@ -0,0 +1,344 @@
+}
+
+// newWebsocketUpgradeConnConfig holds the configuration for creating a new
+// WebsocketUpgradeConn. It includes the context, connection, logger, handshake
+// information, and the type of connection (server or client).
+type newWebsocketUpgradeConnConfig struct {
+	ctx      context.Context
+	conn     net.Conn
+	logger   *slog.Logger
+	hs       ws.Handshake
+	connType connectionType
+}
+
+// newWebsocketUpgradeConn creates a new WebsocketUpgradeConn instance.
+// It initializes the connection with the provided configuration, including
+// the context, connection, logger, and handshake information.
+func newWebsocketUpgradeConn(cfg newWebsocketUpgradeConnConfig) *WebsocketUpgradeConn {
+	return &WebsocketUpgradeConn{
+		logContext:            cfg.ctx,
+		logger:                cfg.logger,
+		conn:                  cfg.conn,
+		supportsCloseProtocol: cfg.hs.Protocol == constants.WebAPIConnUpgradeProtocolWebSocketClose,
+		connType:              cfg.connType,
+		protocol:              cfg.hs.Protocol,
+	}
+}
+
+func (c *WebsocketUpgradeConn) NetConn() net.Conn {
+	return c.conn
+}
+
+func (c *WebsocketUpgradeConn) Read(b []byte) (int, error) {
+	c.readMutex.Lock()
+	defer c.readMutex.Unlock()
+
+	n, err := c.readLocked(b)
+
+	// Timeout errors can be temporary. For example, when this connection is
+	// passed to the kube TLS server, it may get "hijacked" again. During the
+	// hijack, the SetReadDeadline is called with a past timepoint to fail this
+	// Read so that the HTTP server's background read can be stopped. In such
+	// cases, return the original net.Error and clear the cached read error.
+	var netError net.Error
+	if errors.As(err, &netError) && netError.Timeout() {
+		c.readError = nil
+		return n, netError
+	}
+	return n, err
+}
+
+func (c *WebsocketUpgradeConn) readLocked(b []byte) (int, error) {
+	if len(c.readBuffer) > 0 {
+		n := copy(b, c.readBuffer)
+		if n < len(c.readBuffer) {
+			c.readBuffer = c.readBuffer[n:]
+		} else {
+			c.readBuffer = nil
+		}
+		return n, nil
+	}
+
+	// Stop reading if any previous read err.
+	if c.readError != nil {
+		return 0, c.readError
+	}
+
+	for {
+		frame, err := ws.ReadFrame(c.conn)
+		if err != nil {
+			c.readError = err
+			return 0, err
+		}
+
+		// All client frames should be masked.
+		if frame.Header.Masked {
+			frame = ws.UnmaskFrame(frame)
+		}
+
+		switch frame.Header.OpCode {
+		case ws.OpClose:
+			// If we receive a close frame, we should respond with a close frame.
+			if c.supportsCloseProtocol {
+				// Run the close protocol only once to avoid sending frames again
+				// when closing the connection.
+				c.once.Do(func() {
+					if err := c.writeFrame(
+						ws.NewCloseFrame(
+							ws.NewCloseFrameBody(ws.StatusNormalClosure, ""),
+						),
+					); err != nil {
+						if !isOkNetworkErrOrTimeout(err) {
+							c.logger.DebugContext(c.logContext, "error writing close frame", "error", err)
+						}
+					}
+				})
+			}
+			c.readError = io.EOF
+			return 0, io.EOF
```

Comment:
nit: should we more information from the close and use that as the error? like https://github.com/gobwas/ws/blob/v1.4.0/read.go#L123 and maybe also log when received non-normal error

Reply:
We can't. We interleave this code with go std code that does the check:

`err == io.EOF` and if we don't return io.EOF it will fail with errors.
Besides, at this point we always return the normal closure because we can't do much if the connection is closed.

this is the same reason that caused your existing implementation to fail a lot with weird errors because we wrapped them with trace